src/lib.rs

```diff
@@ -0,0 +1,117 @@
+
+pub mod error;
+pub mod nix;
+
+fn process(
+    collected_paths: &Arc<Mutex<std::collections::HashSet<String>>>,
+    flake_ref: &str,
+    system: Option<&str>,
+    attribute_path: &str,
+    offline: &bool,
+    lib: &nix::lib::Lib,
+) -> Vec<DerivationDescription> {
+    log::debug!("Processing derivation: {:?}", attribute_path);
+
+    // call describe_derivation to get the derivation description
+    let description =
+        nix::describe_derivation(flake_ref, system, attribute_path, offline, lib).unwrap();
+
+    // use par_iter to call process on all children of this derivation
+    let children: Vec<DerivationDescription> = description
+        .build_inputs
+        .par_iter()
+        .map(|build_input| {
+            // check if the build_input has already be processed
+            let done = {
+                let mut collected_paths = collected_paths.lock().unwrap();
+                match &build_input.output_path {
+                    Some(output_path) => !collected_paths.insert(output_path.clone()),
+                    None => false,
+                }
+            };
+
+            if done {
+                log::debug!(
+                    "Skipping already processed derivation: {}",
+                    build_input.attribute_path
+                );
+                Vec::new()
+            } else {
+                process(
+                    collected_paths,
+                    flake_ref,
+                    system,
+                    &build_input.attribute_path,
+                    offline,
+                    lib,
+                )
+            }
+        })
+        .flatten()
+        .collect();
+
+    // combine the children and this derivation into a single Vec
+    let mut descriptions = children;
+    descriptions.push(description);
+    descriptions
+}
+
+pub fn nixtract(
+    flake_ref: impl AsRef<str>,
+    system: Option<impl AsRef<str>>,
+    attribute_path: Option<impl AsRef<str>>,
+    offline: &bool,
+) -> Result<Vec<DerivationDescription>> {
+    let flake_ref = flake_ref.as_ref();
+    // Convert system to a Option<&str>
+    let system = system.as_ref().map(AsRef::as_ref);
+    let attribute_path = attribute_path.as_ref().map(AsRef::as_ref);
+
+    // Writes the `lib.nix` file to the tempdir and stores its path
+    let lib = nix::lib::Lib::new()?;
+
+    log::info!(
+        "Starting nixtract with flake_ref: {}, system: {}, attribute_path: {:?}",
+        flake_ref,
+        system
+            .map(AsRef::as_ref)
+            .unwrap_or("builtins.currentSystem"),
+        attribute_path.map(AsRef::as_ref).unwrap_or("")
+    );
+
+    let collected_paths: Arc<Mutex<std::collections::HashSet<String>>> =
+        Arc::new(Mutex::new(std::collections::HashSet::new()));
+
+    // call find_attribute_paths to get the initial set of derivations
+    let attribute_paths =
+        nix::find_attribute_paths(flake_ref, system, attribute_path, offline, &lib)?;
```

Comment:
Would it be possible, like the current (Python) implementation, to start describing derivations before this eval finishes (as it can be pretty long)?

Reply:
Sure, it would be possible. But this current approach is nice and simple, and `nixtract-rs` will mainly be used as a library anyway.